pkgs/cli_util/lib/src/base_directories.dart

```diff
@@ -0,0 +1,274 @@
+  ///
+  /// The directory won't be created, the method merely returns the recommended
+  /// location.
+  ///
+  /// On some platforms, this path may overlap with [cacheHome], and
+  /// [runtimeHome].
+  ///
+  /// Throws an [EnvironmentNotFoundException] if the necessary environment
+  /// variables are undefined.
+  String stateHome() =>
+      path.join(_baseDirectory(_XdgBaseDirectoryType.state), tool);
+
+  String _baseDirectory(_XdgBaseDirectoryType dir) {
+    if (Platform.isWindows) {
+      return _baseDirectoryWindows(dir);
+    }
+    if (Platform.isMacOS) {
+      return _baseDirectoryMacOs(dir);
+    }
+    if (Platform.isLinux) {
+      return _baseDirectoryLinux(dir);
+    }
+    throw UnsupportedError('Unsupported platform: ${Platform.operatingSystem}');
+  }
+
+  String _baseDirectoryWindows(_XdgBaseDirectoryType dir) => switch (dir) {
+        _XdgBaseDirectoryType.config ||
+        _XdgBaseDirectoryType.data =>
+          _requireEnv('APPDATA'),
+        _XdgBaseDirectoryType.cache ||
+        _XdgBaseDirectoryType.runtime ||
+        _XdgBaseDirectoryType.state =>
+          _requireEnv('LOCALAPPDATA'),
+      };
+
+  String _baseDirectoryMacOs(_XdgBaseDirectoryType dir) => switch (dir) {
+        _XdgBaseDirectoryType.config =>
+          path.join(_home, 'Library', 'Preferences'),
+        _XdgBaseDirectoryType.data ||
+        _XdgBaseDirectoryType.state =>
+          path.join(_home, 'Library', 'Application Support'),
+        _XdgBaseDirectoryType.cache => path.join(_home, 'Library', 'Caches'),
+        _XdgBaseDirectoryType.runtime =>
+          // https://stackoverflow.com/a/76799489
+          path.join(_home, 'Library', 'Caches', 'TemporaryItems'),
+      };
+
+  String _baseDirectoryLinux(_XdgBaseDirectoryType dir) {
+    if (Platform.isLinux) {
+      final xdgEnv = switch (dir) {
+        _XdgBaseDirectoryType.config => 'XDG_CONFIG_HOME',
+        _XdgBaseDirectoryType.data => 'XDG_DATA_HOME',
+        _XdgBaseDirectoryType.state => 'XDG_STATE_HOME',
+        _XdgBaseDirectoryType.cache => 'XDG_CACHE_HOME',
+        // Note, DIR instead of HOME.
+        _XdgBaseDirectoryType.runtime => 'XDG_RUNTIME_DIR',
+      };
+      final envVar = environment[xdgEnv];
+      if (envVar != null) {
+        return envVar;
+      }
+    }
+
+    switch (dir) {
+      case _XdgBaseDirectoryType.runtime:
+      // Applications should print a fallback message ideally.
+      case _XdgBaseDirectoryType.cache:
+        return path.join(_home, '.cache');
+      case _XdgBaseDirectoryType.config:
+        return path.join(_home, '.config');
+      case _XdgBaseDirectoryType.data:
+        return path.join(_home, '.local', 'share');
+      case _XdgBaseDirectoryType.state:
+        return path.join(_home, '.local', 'state');
+    }
+  }
+
+  String get _home => _requireEnv('HOME');
+
+  String _requireEnv(String name) =>
+      environment[name] ?? (throw EnvironmentNotFoundException(name));
+}
+
+enum _XdgBaseDirectoryType {
```

Comment:
It's not just XDG, it's for every platform, also the ones that don't use XDG.
I'd use `Kind` instead of `Type` (because `Type` is overloaded and confusing - a type that ends with `Type` is ... redundant? Or it means something else with "Type", but then that isn't clear.)

Or just drop it:
```dart
enum _BaseDirectory {
}
```